### DISCUSSION TITLE
关于`@formily/react`以及`formily/reactive-react`的对等依赖疑问

### DISCUSSION THREAD

**Discussion by lzm0x219** [alibaba/formily#2616] [Q&A 2.x] [answered]
`@formily/react`以及`formily/reactive-react`v2版本都有一个对等依赖：`react-is`。有点没明白这个是需要用户主动安装的吗？

**Comment by janryWang** [comment, marked as answer]
因为内部用了react-is中的isValidElement

**Reply by lzm0x219** [reply]
那为什么会是对等依赖？不应该是生产依赖吗？

**Reply by lzm0x219** [reply]
感觉大多数用户不会消费`react-is`依赖，所以有点没明白把它放到对等依赖的目的是。。？

**Reply by janryWang** [reply]
不太理解，什么是对等依赖，什么是生产依赖，你是说不应该作为peerDependencies？

**Reply by lzm0x219** [reply]
对的

**Reply by lzm0x219** [reply]
peerDependencies就是对等依赖啊。放在peerDependencies不太合适

**Reply by lzm0x219** [reply]
我理解的：生产依赖就是源码中引入的依赖，对等依赖则是库源码中需要，且目标用户项目源码也需要用到的依赖。

**Reply by janryWang** [reply]
不用peerDeps会导致重复打包，比如用户用React16，formily内部安装React17的，那就会容易重复打包了

**Reply by lzm0x219** [reply]
我是觉得`react-is`这个依赖应该放在生产依赖里，或者是可选对等依赖，因为它并不是像`react`一样是项目强依赖，

**Reply by lzm0x219** [reply]
不应该丢给用户去强制消费这个依赖

**Reply by janryWang** [reply]
首先，重复打包的问题，怎么解决？麻烦给个方案，而不是说，这个不应该那个不应该，咱们要针对具体问题看问题

**Reply by lzm0x219** [reply]
明白你的意思了，为了处理重复打包所以才丢到对等去的

**Reply by lzm0x219** [reply]
> 首先，重复打包的问题，怎么解决？麻烦给个方案，而不是说，这个不应该那个不应该，咱们要针对具体问题看问题

丢到shared包统一管理，子包都从shared包导入

**Reply by janryWang** [reply]
没必要呀，放shared里管理，shared里肯定也得用peerDeps

**Reply by lzm0x219** [reply]
shared里设为可选对等就行了

**Reply by janryWang** [reply]
什么是可选对等，具体怎么写

**Reply by lzm0x219** [reply]
> 什么是可选对等，具体怎么写

![image](https://user-images.githubusercontent.com/50414099/145747081-ed555370-a85c-4d77-9e8c-2fab3de924e3.png)

**Reply by janryWang** [reply]
那直接给reactive-react和react两个包写可选对等依赖就行了吧，也不用挪到shared里，shared是不依赖react的

**Reply by lzm0x219** [reply]
理论上是这样的，挪到shared里也只是为了统一同级包引入的版本号一致。但是仍然需要测试，并且我看MUI团队并没有对此做特殊处理，[mui-material-next](https://github.com/mui-org/material-ui/blob/master/packages/mui-material-next/package.json), [mui-base](https://github.com/mui-org/material-ui/blob/master/packages/mui-base/package.json) 他们的依赖关系与formily相似

**Reply by janryWang** [reply]
嗯嗯，那来个PR吧，可选对等应该能解决问题

**Reply by lzm0x219** [reply]
可以
